daemon/attach.go

```diff
@@ -52,10 +52,27 @@ func (daemon *Daemon) ContainerAttach(prefixOrName string, c *backend.ContainerA
 	ctr.StreamConfig.AttachStreams(&cfg)
 
 	multiplexed := !ctr.Config.Tty && c.MuxStreams
-	inStream, outStream, errStream, err := c.GetStreams(multiplexed)
+
+	clientCtx, closeNotify := context.WithCancel(context.Background())
+	defer closeNotify()
+	go func() {
+		<-clientCtx.Done()
+		// The client has disconnected
+		// In this case we need to close the container's output streams so that the goroutines used to copy
+		// to the client streams are unblocked and can exit.
+		if cfg.CStdout != nil {
+			cfg.CStdout.Close()
+		}
+		if cfg.CStderr != nil {
+			cfg.CStderr.Close()
+		}
+	}()
+
+	inStream, outStream, errStream, err := c.GetStreams(multiplexed, closeNotify)
```

Comment:
Can we get a FIXME here that notes we should properly plumb contexts instead of passing a notification callback?

Reply:
I'm not sure what you mean here.
At best we would need to pass a context back rather than a passing in `closeNotify`.

e.g. `ctx, inStream, outStream, errStream, err := c.GetStreams(ctx, multiplexed)`

`c` here is not the container, its the attach config (and I suppose poorly referenced).

Reply:
Not sure if this is the actual confusing here, but I went ahead and renamed `c` to `req` because `c` is usually `container`, and there is also `ctr` and `cfg`... so lots of things to be confused on when reading the code in general.